.github/workflows/continuous-integration.yml

```diff
@@ -24,12 +24,21 @@ jobs:
         run: bash scripts/ci/inspect_binary_file_sizes.sh ${{ github.sha }} ${{ github.sha }}
 
       - name: set up Node.js
+        id: setup-node
         uses: actions/setup-node@v3
         with:
           cache: npm
           node-version: 16
 
+      - name: load node_modules from cache
+        id: cache-node-modules
+        uses: actions/cache@v3
+        with:
+          path: node_modules
+          key: node-modules-${{ hashFiles('package-lock.json') }}-${{ steps.setup-node.outputs.node-version }}
+
       - name: install remark
+        if: steps.cache-node-modules.outputs.cache-hit != 'true'
```

Comment:
Is it not quite safe to leave this in place? Is there much of a saving from skipping this command run?

Reply:
it's basically the gist of the whole PR. not skipping it results in `npm ci` installing it all from scratch (see timings from https://github.com/TicClick/osu-wiki/pull/11)

by the way, why is it not safe? reading cache is scoped to the feature branch and its base for read, and writing to the branch itself, so you can't affect someone else's PR by doing changes in yours (unless I misunderstand your question)

Reply:
for instance, in the case the cache is in a bad state (ie. a "bpm download failure" gets cached somehow) there is no post step to verify and recover

Reply:
`npm ci` is the only thing we use that writes to `node_modules`. if it fails, the whole workflow fails earlier than cache creation takes place (it's one of the last steps).

I can look into verifying the cache on hit but I think we're still good

Reply:
Can try and see how it plays out. I just know how much of a PITA cache can be if you cache wrongly.

Reply:
done some digging:
- apparently `npm c(lean)i(nstall)` does its own integrity checks
  - removing `node_modules` is also the first thing it does, even with `--dry-run`
- running `npm install --no-audit --dry-run` validates nothing, and `npm ci --no-audit` provides only a marginal gain
- you can't quickly verify integrity of all installed packages
- at the same time, GitLab CI themselves mention `node_modules` as one of folders potentially subject to caching in their workflows: https://docs.gitlab.com/ee/ci/caching/#cache-nodejs-dependencies

I am convinced that if for whatever reason the cache will end in a bad state, linter will also fail and we'll take note of that. as an extra precaution against it being created already malformed, I can add `hashFiles(package.json)`

Reply:
I used `npm ci` instead of `npm i` originally since the idea was to cache `~/.npm`, which is npm's internal cache for package downloads, then rebuild node_modules every run. `npm i` can be used if you *want* to keep previous node_modules

the difference between caching .npm and node_modules for a one-project environment like this comes down to whether it's faster to unpack from a smaller cache (.npm) or use a larger cache (node_modules). haven't tested that myself, I just used the github built-in option (.npm)